texprintf.3
.TH texprintf 1.0.0 "13 Jan 2019"
.SH NAME
texprintf, stexprintf, ftexprintf, texlistsymbols, texboxtree \- Extended formatted output functions
.SH SYNOPSIS
.nf
.B #include <texprintf.h>
.PP
.BI "int texprintf(const char *format, ...);"
.BI "char * stexprintf(const char *format, ...);"
.BI "int ftexprintf(const char *format, ...);"
.BI "void texlistsymbols();"
.BI "void texboxtree(const char *format, ...);"
.BI "void texlistsymbols();"
.BI "void texerrors();"
.fi
.PP
In adition to the above functions <texprintf.h> defines several variables:
.nf
.PP
.BI "int TEXPRINTF_LW;"
.BI "char * TEXPRINTF_FONT;"
.BI "int TEXPRINTF_ERR;"
.fi
.PP

.SH DESCRIPTION
The functions in the texprintf family provide formatted output like the printf family of functions extended
with a tex-like syntax. The functions generate UTF-8 encoded strings and assume that monospace fonts are used
(also for symbols, beware of non-monospace fallback fonts).

.br
Please refer to the utftex man page for details on the tex-like syntax provided by these functions. Refer to
the man page of printf for a description of the format.

.br
Upon successful return the functions texprintf, and ftexprintf return the number of characters
printed (excluding the null byte used to end output to strings). The stexprintf function
returns an allocated string containing the result. 

.br
With the integer TEXPRINTF_LW, the line width may be limited. Per default the line width is not limited 
(TEXPRINTF_LW=0). If TEXPRINTF_LW is set to a positive integer, the texprintf functions will 
.B try 
to limit the line width to this number of characters, i.e., tt tries to break up the equation 
over more than one line. To this end the texprinf engine must find suitable places to break the equation.
As it is not guaranteed a suitable location is found, the line width is 
.B not
guaranteed.

.br
The string TEXPRINTF_FONT can be used to specify the default font. The available fonts are
.nf
.BI "mathsfbfit" 
.BI "mathsfbf" 
.BI "mathfrak" 
.BI "mathbfit" 
.BI "mathsfit" 
.BI "mathcal" 
.BI "mathscr" 
.BI "mathbf" 
.BI "mathbb" 
.BI "mathsf" 
.BI "mathtt" 
.BI "mathnormal"
.BI "text (default)"
.fi
.br
The integer TEXPRINTF_ERR is an error flag. If its value is non zero after a call to any of the texprintf functions,
an error has occurred in parsing the input. A call to texerrors() will print the corresponding error messages. 
A call to texerrors() will print the related error messages.
.br
The function texlistsymbols() prints all symbols the texprintf system knows. 
.br
The texboxtree(const char *format, ...) is for debugginbg purposes. It takes the same arguments as texprintf, however,
its output is a description of the resulting internal box datastructures.  


.SH SEE ALSO
printf(3), utftex(1)



src/texprintf.c
#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <stdarg.h>
#include "boxes.h"
#include "drawbox.h"
#include "parser.h"
#include "lexer.h"
#include "stringutils.h"
#include "error.h"

int TEXPRINTF_LW=0;
char * TEXPRINTF_FONT="text";
int TEXPRINTF_FCW=2;
int TEXPRINTF_WCW=2;

int TEXPRINTF_ERR=0;

int texprintf(const char *format, ...)
{
   	va_list ap;
   	char *str;
   	int Na=255, np;
	box root;
   	
	ResetErrors();
	FCSPACES=TEXPRINTF_FCW;
	WCSPACES=TEXPRINTF_WCW;
	if ((WCSPACES<1)||(WCSPACES>2))
	{
		fprintf(stderr,"Error: TEXPRINTF_WCW out of range, wide characters can occupy either 1 or 2 character spaces\n");
		exit(1);
	}
	if ((FCSPACES<1)||(FCSPACES>2))
	{
		fprintf(stderr,"Error: TEXPRINTF_FCW out of range, wide characters can occupy either 1 or 2 character spaces\n");
		exit(1);
	}
	
   	str=malloc(Na*sizeof(char));
   	va_start (ap, format);
   	np=vsnprintf(str, Na*sizeof(char), format, ap);
   	while (np>=Na)
   	{
		Na+=255;
		str=realloc(str, Na*sizeof(char));
		va_start (ap, format);
		np=vsnprintf(str, Na*sizeof(char), format, ap);
	}
	
	root=ParseString(str, TEXPRINTF_LW, TEXPRINTF_FONT);
	BoxPos(&root);
	np=PrintBox(&root);
	FreeBox(&root);
	free(str);
	TEXPRINTF_ERR=ERRORSTATE;
	return np;
}


char * stexprintf(const char *format, ...)
{
   	va_list ap;
   	char *res;
   	char *str;
   	int Na=255, np;
	box root;
   	
	ResetErrors();
	FCSPACES=TEXPRINTF_FCW;
	WCSPACES=TEXPRINTF_WCW;
	if ((WCSPACES<1)||(WCSPACES>2))
	{
		fprintf(stderr,"Error: TEXPRINTF_WCW out of range, wide characters can occupy either 1 or 2 character spaces\n");
		exit(1);
	}
	if ((FCSPACES<1)||(FCSPACES>2))
	{
		fprintf(stderr,"Error: TEXPRINTF_FCW out of range, wide characters can occupy either 1 or 2 character spaces\n");
		exit(1);
	}
   	
   	str=malloc(Na*sizeof(char));
   	va_start (ap, format);
   	np=vsnprintf(str, Na*sizeof(char), format, ap);
   	while (np>=Na)
   	{
		Na+=255;
		str=realloc(str, Na*sizeof(char));
		va_start (ap, format);
		np=vsnprintf(str, Na*sizeof(char), format, ap);
	}
	
	root=ParseString(str, TEXPRINTF_LW, TEXPRINTF_FONT);
	BoxPos(&root);
	res=DrawBox(&root);
	FreeBox(&root);
	free(str);	
	TEXPRINTF_ERR=ERRORSTATE;

	return res;
}

int ftexprintf(FILE *f, const char *format, ...)
{
   	va_list ap;
   	char *res;
   	char *str;
   	int Na=255, np;
	box root;
   	
	ResetErrors();
	FCSPACES=TEXPRINTF_FCW;
	WCSPACES=TEXPRINTF_WCW;
	if ((WCSPACES<1)||(WCSPACES>2))
	{
		fprintf(stderr,"Error: TEXPRINTF_WCW out of range, wide characters can occupy either 1 or 2 character spaces\n");
		exit(1);
	}
	if ((FCSPACES<1)||(FCSPACES>2))
	{
		fprintf(stderr,"Error: TEXPRINTF_FCW out of range, wide characters can occupy either 1 or 2 character spaces\n");
		exit(1);
	}
   	
   	str=malloc(Na*sizeof(char));
   	va_start (ap, format);
   	np=vsnprintf(str, Na*sizeof(char), format, ap);
   	while (np>=Na)
   	{
		Na+=255;
		str=realloc(str, Na*sizeof(char));
		va_start (ap, format);
		np=vsnprintf(str, Na*sizeof(char), format, ap);
	}
	
	root=ParseString(str, TEXPRINTF_LW, TEXPRINTF_FONT);
	BoxPos(&root);
	res=DrawBox(&root);
	FreeBox(&root);
	free(str);
	np=strlen(res);
	fprintf(f,"%s",res);
	free(res);	
	TEXPRINTF_ERR=ERRORSTATE;

	return np;
}

void texboxtree(const char *format, ...)
{
   	va_list ap;
   	char *str;
   	int Na=255, np;
	box root;
   	
	ResetErrors();
	FCSPACES=TEXPRINTF_FCW;
	WCSPACES=TEXPRINTF_WCW;
	if ((WCSPACES<1)||(WCSPACES>2))
	{
		fprintf(stderr,"Error: TEXPRINTF_WCW out of range, wide characters can occupy either 1 or 2 character spaces\n");
		exit(1);
	}
	if ((FCSPACES<1)||(FCSPACES>2))
	{
		fprintf(stderr,"Error: TEXPRINTF_FCW out of range, wide characters can occupy either 1 or 2 character spaces\n");
		exit(1);
	}
	
   	str=malloc(Na*sizeof(char));
   	va_start (ap, format);
   	np=vsnprintf(str, Na*sizeof(char), format, ap);
   	while (np>=Na)
   	{
		Na+=255;
		str=realloc(str, Na*sizeof(char));
		va_start (ap, format);
		np=vsnprintf(str, Na*sizeof(char), format, ap);
	}
	
	root=ParseString(str, TEXPRINTF_LW, TEXPRINTF_FONT);
	BoxPos(&root);
	DrawBoxTree(&root);
	FreeBox(&root);
	free(str);
	TEXPRINTF_ERR=ERRORSTATE;

}

void texlistsymbols()
{
	ListSymbols();								/* link to lexer.o */
}

void texerrors()
{
	ERRORSTATE=TEXPRINTF_ERR;
	E_Messages();						/* link to error.o */
	ResetErrors();
}

void SetStyleASCII()
{
	style=&STYLE_ASC;
}
void SetStyleUNICODE()
{
	style=&STYLE_UNI;
}
